File: csharp/Policy.cs
namespace Jev;

public enum Tri
{
    False,
    True,
    Uncertain,
}

public sealed class TriConverter : JsonConverter<Tri>
{
    public override Tri Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) =>
        throw new NotSupportedException();

    public override void Write(Utf8JsonWriter writer, Tri value, JsonSerializerOptions options)
    {
        switch (value)
        {
            case Tri.True:
                writer.WriteBooleanValue(true);
                break;
            case Tri.False:
                writer.WriteBooleanValue(false);
                break;
            default:
                writer.WriteStringValue("uncertain");
                break;
        }
    }
}

public sealed class Policy
{
    public double SwitchConfidence { get; set; } = 0.72;
    public double InterruptAt { get; set; } = 0.75;
    public double UncertainMargin { get; set; } = 0.12;
}

public sealed record Decision(string Disposition, string Executing, string Because);

public sealed class TargetCandidate
{
    public string Id { get; set; } = "";
    public string Relation { get; set; } = "";
    public double? Distance { get; set; }
    public double? Health { get; set; }
    public string Kind { get; set; } = "";
}

public sealed class Bound
{
    public string? TargetId { get; set; }
    public string TargetSource { get; set; } = "none";
}

public sealed class ScoreRead
{
    public double Score { get; set; }
    public int Level { get; set; }
    public string Label { get; set; } = "";
}

public static class PolicyEngine
{
    public static Policy Resolve(JsonObject? input)
    {
        var policy = new Policy();
        if (input is null) return policy;
        foreach (var key in new[] { "switchConfidence", "interruptAt", "uncertainMargin" })
        {
            if (!input.TryGetPropertyValue(key, out var raw) || raw is null) continue;
            if (!JsonUtil.TryFloat(raw, out var value) || !double.IsFinite(value) || value < 0 || value > 1)
                throw Errors.Bad(400, $"policy.{key} 必须是 0 到 1 之间的数字");
            switch (key)
            {
                case "switchConfidence": policy.SwitchConfidence = value; break;
                case "interruptAt": policy.InterruptAt = value; break;
                default: policy.UncertainMargin = value; break;
            }
        }
        return policy;
    }

    public static Tri ClassifyBoolean(double? probability, double margin)
    {
        if (probability is not double p || !double.IsFinite(p)) return Tri.Uncertain;
        if (p >= 0.5 + margin) return Tri.True;
        if (p <= 0.5 - margin) return Tri.False;
        return Tri.Uncertain;
    }

    public static Tri ClassifyInterrupt(double? probability, double interruptAt)
    {
        if (probability is not double p || !double.IsFinite(p)) return Tri.Uncertain;
        if (p >= interruptAt) return Tri.True;
        if (p <= 1 - interruptAt) return Tri.False;
        return Tri.Uncertain;
    }

    public static Decision DecideDisposition(string current, string suggested, double? confidence, Tri interrupt, string? holdKey, double switchConfidence)
    {
        current = current.Trim();
        var expanded = holdKey == suggested && current.Length > 0 ? current : suggested;
        if (current.Length == 0) return new Decision("switch", expanded, "first-decision");
        if (expanded == current) return new Decision("continue", current, "still-fitting");
        if (interrupt == Tri.True) return new Decision("switch", expanded, "interrupt");
        if (confidence is null || confidence >= switchConfidence) return new Decision("switch", expanded, "confident");
        return new Decision("hold", current, "hysteresis");
    }

    public static bool TacticNeedsTarget(string tactic) => tactic is not ("hold" or "patrol" or "hide" or "flee");

    public static Bound BindTarget(string tactic, string modelTargetId, IReadOnlyList<TargetCandidate> roster)
    {
        if (!TacticNeedsTarget(tactic)) return new Bound { TargetSource = "none" };
        if (modelTargetId is not ("none" or "") && roster.Any(entry => entry.Id == modelTargetId))
            return new Bound { TargetId = modelTargetId, TargetSource = "model" };
        var fallback = FallbackTarget(tactic, roster);
        if (fallback is not null) return new Bound { TargetId = fallback, TargetSource = "geometric-fallback" };
        return new Bound { TargetSource = "none" };
    }

    public static ScoreRead ScoreBand(double score, IReadOnlyList<string> labels)
    {
        var max = Math.Max(0, labels.Count - 1);
        var clamped = Math.Clamp(score, 0, max);
        var level = (int)Math.Clamp(Math.Round(clamped), 0, max);
        return new ScoreRead
        {
            Score = Math.Round(clamped * 100) / 100,
            Level = level,
            Label = level < labels.Count ? labels[level] : "unknown",
        };
    }

    private static string? FallbackTarget(string tactic, IReadOnlyList<TargetCandidate> roster)
    {
        var ranked = roster.OrderBy(DistOrFar).ToList();
        if (tactic == "assist")
        {
            return ranked.Where(entry => entry.Relation == "ally")
                .OrderBy(entry => entry.Health ?? 1)
                .ThenBy(DistOrFar)
                .Select(entry => entry.Id)
                .FirstOrDefault();
        }
        if (tactic is "investigate" or "interact")
        {
            return ranked.FirstOrDefault(entry => entry.Kind is "interest" or "hazard" or "prop")?.Id
                ?? ranked.FirstOrDefault()?.Id;
        }
        return ranked.FirstOrDefault(entry => entry.Relation == "enemy")?.Id ?? ranked.FirstOrDefault()?.Id;
    }

    private static double DistOrFar(TargetCandidate item) => item.Distance ?? 1e9;
}

public static class JsonUtil
{
    public static readonly JsonSerializerOptions Options = CreateOptions();

    public static bool TryFloat(JsonNode? node, out double value)
    {
        value = 0;
        if (node is not JsonValue raw) return false;
        if (raw.TryGetValue<double>(out value)) return true;
        if (raw.TryGetValue<int>(out var i)) { value = i; return true; }
        if (raw.TryGetValue<long>(out var l)) { value = l; return true; }
        if (raw.TryGetValue<float>(out var f)) { value = f; return true; }
        return false;
    }

    private static JsonSerializerOptions CreateOptions()
    {
        var options = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
            Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
        };
        options.Converters.Add(new TriConverter());
        return options;
    }
}

File: src/policy.ts
import { IntuitionError } from './errors'
import type { Because, PolicyConfig, Relation, Tri } from './types.js'

export interface Policy {
  switchConfidence: number
  interruptAt: number
  uncertainMargin: number
}

export const DEFAULT_POLICY: Policy = {
  switchConfidence: 0.72,
  interruptAt: 0.75,
  uncertainMargin: 0.12,
}

export interface Decision {
  disposition: 'continue' | 'switch' | 'hold'
  executing: string
  because: Because
}

export interface TargetCandidate {
  id: string
  relation: Relation
  distance: number | null
  health: number | null
  kind: string
}

const NO_TARGET = new Set(['hold', 'patrol', 'hide', 'flee'])

export function resolvePolicy(input?: PolicyConfig): Policy {
  const policy: Policy = { ...DEFAULT_POLICY }
  if (!input) return policy
  for (const key of ['switchConfidence', 'interruptAt', 'uncertainMargin'] as const) {
    const value = input[key]
    if (value == null) continue
    if (typeof value !== 'number' || Number.isNaN(value) || value < 0 || value > 1) {
      throw new IntuitionError(400, `policy.${key} 必须是 0 到 1 之间的数字`)
    }
    policy[key] = value
  }
  return policy
}

export function classifyBoolean(probability: number | null, margin: number): Tri {
  if (probability == null || Number.isNaN(probability)) return 'uncertain'
  if (probability >= 0.5 + margin) return true
  if (probability <= 0.5 - margin) return false
  return 'uncertain'
}

export function classifyInterrupt(probability: number | null, interruptAt: number): Tri {
  if (probability == null || Number.isNaN(probability)) return 'uncertain'
  if (probability >= interruptAt) return true
  if (probability <= 1 - interruptAt) return false
  return 'uncertain'
}

/**
 * Jev 的每个问题互相独立，所以代码只做三件事：
 * 展开 hold、用 interrupt 放行一次切换、置信度不够就保持当前战术。
 * interrupt 不能推翻模型明确选中的 hold。
 */
export function decideDisposition(args: {
  current: string | null | undefined
  suggested: string
  confidence: number | null
  interrupt: Tri
  holdKey: string | null
  switchConfidence: number
}): Decision {
  const current = args.current?.trim() || null
  const expanded =
    args.holdKey != null && args.suggested === args.holdKey && current ? current : args.suggested

  if (!current) {
    return { disposition: 'switch', executing: expanded, because: 'first-decision' }
  }
  if (expanded === current) {
    return { disposition: 'continue', executing: current, because: 'still-fitting' }
  }
  if (args.interrupt === true) {
    return { disposition: 'switch', executing: expanded, because: 'interrupt' }
  }
  if (args.confidence == null || args.confidence >= args.switchConfidence) {
    return { disposition: 'switch', executing: expanded, because: 'confident' }
  }
  return { disposition: 'hold', executing: current, because: 'hysteresis' }
}

export function tacticNeedsTarget(tactic: string): boolean {
  return !NO_TARGET.has(tactic)
}

export function bindTarget(
  tactic: string,
  modelTargetId: string,
  roster: TargetCandidate[],
): { targetId: string | null; targetSource: 'model' | 'none' | 'geometric-fallback' } {
  if (!tacticNeedsTarget(tactic)) return { targetId: null, targetSource: 'none' }
  if (modelTargetId !== 'none' && roster.some((entry) => entry.id === modelTargetId)) {
    return { targetId: modelTargetId, targetSource: 'model' }
  }
  const fallback = fallbackTarget(tactic, roster)
  if (fallback) return { targetId: fallback, targetSource: 'geometric-fallback' }
  return { targetId: null, targetSource: 'none' }
}

export function scoreBand(
  score: number,
  labels: readonly string[],
): { score: number; level: number; label: string } {
  const max = Math.max(0, labels.length - 1)
  const clamped = Math.min(max, Math.max(0, score))
  const level = Math.min(max, Math.max(0, Math.round(clamped)))
  return { score: round2(clamped), level, label: labels[level] ?? labels[0] ?? 'unknown' }
}

function fallbackTarget(tactic: string, roster: TargetCandidate[]): string | null {
  const ranked = [...roster].sort((a, b) => (a.distance ?? 1e9) - (b.distance ?? 1e9))
  if (tactic === 'assist') {
    const allies = ranked
      .filter((entry) => entry.relation === 'ally')
      .sort(
        (a, b) =>
          (a.health ?? 1) - (b.health ?? 1) || (a.distance ?? 1e9) - (b.distance ?? 1e9),
      )
    return allies[0]?.id ?? null
  }
  if (tactic === 'investigate' || tactic === 'interact') {
    const prop = ranked.find(
      (entry) => entry.kind === 'interest' || entry.kind === 'hazard' || entry.kind === 'prop',
    )
    return prop?.id ?? ranked[0]?.id ?? null
  }
  return ranked.find((entry) => entry.relation === 'enemy')?.id ?? ranked[0]?.id ?? null
}

function round2(n: number): number {
  return Math.round(n * 100) / 100
}
